src/react/utils.js

```diff
@@ -555,14 +557,64 @@ export function deleteRefAndKeyFromProps(realm: Realm, props: ObjectValue | Abst
   deleteProperty(props, "key");
 }
 
-export function objectHasNoPartialKeyAndRef(
+export function resetRefAndKeyFromProps(realm: Realm, props: Value): void {
+  if (props instanceof NullValue) {
+    return;
+  }
+  invariant(props instanceof ObjectValue || props instanceof AbstractObjectValue);
+  if (props instanceof AbstractObjectValue) {
+    if (props.values.isTop()) {
+      let template = Create.ObjectCreate(realm, realm.intrinsics.ObjectPrototype);
+      template.makePartial();
+      template.makeSimple();
+      props.values = new ValuesDomain(new Set([template]));
```

Comment:
I don't understand what this is doing. Can you explain?

Reply:
We create a template object that can contain values (in this case, `key` and `ref` and make the bindings with undefined descriptors). We then need to attach it to an abstract object value so we need to ensure it has a values domain so we can add it to its Set.

Reply:
You’re describing the mechanics of how to construct Prepack objects. But I’m asking about the higher level picture of what you’re trying to do and why it is the correct thing to do.

Reply:
We need to create a template to store the fact we want to make `key` and `ref`. In order to do that we need to create a backing template so we can `Set` values. You can't `Set` values on an abstract object value that is "top" (one without a template).